test/gtest/gpu_nchw_nhwc_transpose.cpp

```diff
@@ -0,0 +1,658 @@
+void cpu_nhwc2nchw(T* dst, T* src, uint64_t N, uint64_t C, uint64_t H, uint64_t W)
+{
+    cpu_ndhwc2ncdhw<T>(dst, src, N, C, 1, H, W);
+}
+
+template <typename T, typename TRANSPOSE_SOL>
+struct cpu_transpose
+{
+};
+
+template <typename T>
+struct cpu_transpose<T, miopen::TransposeSolutionDefault2Nhwc>
+{
+    static void run(T* dst, T* src, uint64_t N, uint64_t C, uint64_t H, uint64_t W)
+    {
+        cpu_nchw2nhwc<T>(dst, src, N, C, H, W);
+    }
+
+    static void run(T* dst, T* src, uint64_t N, uint64_t C, uint64_t D, uint64_t H, uint64_t W)
+    {
+        cpu_ncdhw2ndhwc<T>(dst, src, N, C, D, H, W);
+    }
+};
+
+template <typename T>
+struct cpu_transpose<T, miopen::TransposeSolutionDefault2Ndhwc>
+{
+    static void run(T* dst, T* src, uint64_t N, uint64_t C, uint64_t D, uint64_t H, uint64_t W)
+    {
+        cpu_ncdhw2ndhwc<T>(dst, src, N, C, D, H, W);
+    }
+};
+
+template <typename T>
+struct cpu_transpose<T, miopen::TransposeSolutionNhwc2Default>
+{
+    static void run(T* dst, T* src, uint64_t N, uint64_t C, uint64_t H, uint64_t W)
+    {
+        cpu_nhwc2nchw<T>(dst, src, N, C, H, W);
+    }
+
+    static void run(T* dst, T* src, uint64_t N, uint64_t C, uint64_t D, uint64_t H, uint64_t W)
+    {
+        cpu_ndhwc2ncdhw<T>(dst, src, N, C, D, H, W);
+    }
+};
+
+template <typename T>
+struct cpu_transpose<T, miopen::TransposeSolutionNdhwc2Default>
+{
+    static void run(T* dst, T* src, uint64_t N, uint64_t C, uint64_t D, uint64_t H, uint64_t W)
+    {
+        cpu_ndhwc2ncdhw<T>(dst, src, N, C, D, H, W);
+    }
+};
+
+template <typename TRANSPOSE_SOL>
+struct transpose_str
+{
+};
+
+template <>
+struct transpose_str<miopen::TransposeSolutionDefault2Nhwc>
+{
+    static std::string get() { return "nchw2nhwc"; }
+};
+
+template <>
+struct transpose_str<miopen::TransposeSolutionNhwc2Default>
+{
+    static std::string get() { return "nhwc2nchw"; }
+};
+
+template <typename T>
+struct to_miopen_data_type
+{
+};
+
+template <>
+struct to_miopen_data_type<float>
+{
+    static miopenDataType_t get() { return miopenFloat; }
+};
+
+template <>
+struct to_miopen_data_type<float16>
+{
+    static miopenDataType_t get() { return miopenHalf; }
+};
+
+template <>
+struct to_miopen_data_type<bfloat16>
+{
+    static miopenDataType_t get() { return miopenHalf; }
+};
+
+// TODO: try to get F8 working
+template <>
+struct to_miopen_data_type<miopen_f8::hip_f8<miopen_f8::hip_f8_type::fp8>>
+{
+    static miopenDataType_t get() { return miopenFloat8; }
+};
+
+template <>
+struct to_miopen_data_type<uint16_t>
+{
+    static miopenDataType_t get() { return miopenHalf; } // we actually didn't calculate 16bit float
+};
+
+template <>
+struct to_miopen_data_type<uint8_t>
+{
+    static miopenDataType_t get() { return miopenInt8; }
+};
+
+static constexpr int RAND_INTEGER_MAX = 120;
+static constexpr int RAND_INTEGER_MIN = -88;
+
+template <typename T>
+void rand_tensor_integer(tensor<T>& t, int max = RAND_INTEGER_MAX, int min = RAND_INTEGER_MIN)
```

Comment:
remove and replace with `tensor.generate()` from `tensor_holder.hpp`.

Reply:
done (in layout_transpose.cpp)